### PR TITLE
Teleport System v2

### PR DESCRIPTION
**Pull request by FejZa**
# XRTK - Mixed Reality Toolkit Pull Request

## Overview

This PR fixes https://github.com/XRTK/XRTK-Core/issues/504 and introduces some new features to the teleport system.
The core change is that the act of "teleporting" is now done by "teleportation handlers". Instead of the teleport system just moving the camera around. Using a new teleport system profile the user can configure a handler implementation to use, this opens up endless possibilities to handle teleportation.

Previously:
- Teleport just moves the camera/playspace to target position instantly. That's it

Now:
- You can either just teleport or have e.g. an animation / sound or whatever play when teleporting. It's all up to how the handler implements it

The PR comes with two default implementations.

### `InstantTeleportHandler`

This is what we had so far in terms of looks and behaviour. The user selects a teleport location and instantly is moved there.

instant-teleport_1

### `FadingTeleportHandler`

This implementation will fade in / out the camera before teleporting, which leads to a more appealing and smooth teleport.

fade-teleport_1

## Changes

- Fixes: https://github.com/XRTK/XRTK-Core/issues/504

## Submodule Changes

- [SDK](https://github.com/XRTK/SDK/pull/213)
- [Examples](https://github.com/XRTK/Examples/pull/19)

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Services/TeleportSystem/MixedRealityTeleportSystem.cs [hunk 1]
```diff
@@ -190,45 +196,10 @@ public void RaiseTeleportCanceled(IMixedRealityPointer pointer, IMixedRealityTel
...
-            var height = targetPosition.y + parentPosition.y;
-            targetPosition -= cameraPosition - parentPosition;
-            targetPosition.y = height;
-            cameraParent.position = targetPosition;
-            cameraParent.RotateAround(cameraPosition, Vector3.up, targetRotation.y - cameraTransform.eulerAngles.y);
-
-            isProcessingTeleportRequest = false;
-
-            // Raise complete event using the pointer and hot spot provided.
-            RaiseTeleportComplete(eventData.Pointer, eventData.HotSpot);
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Runtime/Services/TeleportSystem/MixedRealityTeleportSystem.cs:231]: I take this all this was moved to the TeleportHandler implementation?

**Comment by StephenHodgson** [comment]
Overall love the work, I think it's great we have a proper profile and handlers for this now.

I guess my only concern is that the default implementation cannot be packaged with the core (similar to the gaze data provider which serves a similar function as it is required for the input system to work)

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Services/TeleportSystem/MixedRealityTeleportSystem.cs (hunk 1 above)
Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Services/TeleportSystem/MixedRealityTeleportSystem.cs:231]: Right

**Comment by FejZa** [comment]
But don't most of the systems have that issue? Spatial Awareness also won't be able to do anything without SDK.

**Comment by FejZa** [comment]
We could have a "fallback default implementation" that kicks in if no handlers is available. As soon as one is registered in the system profile it takes over?

**Comment by StephenHodgson** [comment]
Yeah but the difference with the Spatial Awareness system is that not all platforms could have data providers.

I think it's more like the default gaze provider.

**Review by StephenHodgson — APPROVED** [review]
Got to try out the demos, great stuff Dino!

**Pull request merged**